app/models/mixins/authentication_mixin.rb

```diff
@@ -2,10 +2,49 @@ module AuthenticationMixin
   extend ActiveSupport::Concern
 
   included do
+    # There are some dirty, dirty Ruby/Rails reasons why this method needs to
+    # exist like this, and I will try and explain:
+    #
+    # First, we need the extra +.where+ here because if this is used in a
+    # nested SELECT statement (aka: virtual_delegate), the +resource_type+
+    # check is dropped on the floor and not included in the subquery.  As a
+    # result, it will just pickup the first record to match the
+    # relationship_id column and that could be associated with any OTHER object
+    # with an Authentication record.
+    #
+    # For example, you get back an authentication record for an EMS record when
+    # you are looking up on for a Host.
+    #
+    # Secondly, the reason this method exists, and is FIRST (prior to the
+    # +has_one+ that uses it) is it needs to be defined prior to the
+    # relationship method that calls it.  Also, it needs to be a method so the
+    # proper local variable, +authentication_mixin_relation+, can be defined
+    # with the class that the +resource_type+ needs to match and remain in
+    # scope for the Proc below.  If it is a class variable or done in other
+    # fashion, it will be overwritten whenever this module is included
+    # elsewhere.
+    #
+    # Finally, the use of a prepared statement is also for some reason required
+    # over the Hash syntax since otherwise the following is ERROR is produced
+    # when doing a nested SELECT:
+    #
+    #     PG::ProtocolViolation: ERROR:  bind message supplies 0 parameters,
+    #     but prepared statement "" requires 1 (ActiveRecord::StatementInvalid)
+    #
+    def self.authentication_status_severity_level_filter
+      # required to be done here so it is in scope of the Proc below
+      authentication_mixin_relation = name
+
+      proc do
+        where('"authentications"."resource_type" = ?', authentication_mixin_relation)
```

Comment:
yea, I really do like the hash instead:

```suggestion
        where(:authentications => {:resource_type => authentication_mixin_relation})
```

I kinda like doing this inlined in the actual method, especially since this is fixing a bug in a component outside miq. with a comment in the code

but not neither of those are my hill

Reply:
Keenan's suggestion looks much more Ruby-ish, while the Nick's looks SQL-ish.  Prefer Ruby ;-)

Reply:
@chessbyte @kbrock I already responded to @Fryguy about this:

https://github.com/ManageIQ/manageiq/pull/20494#discussion_r478621643

See the error that I commented about two:

https://github.com/ManageIQ/manageiq/pull/20494/files#diff-f9cace1fe4fdcec66f962b23f2110fd6R31-R32

This is when I used the Hash syntax... I don't think nesting the hash will help in this case, but I can try in a bit.

Reply:
Okay, did want to "check myself" before I decided to be definitive about this, because I originally just tested with:

```ruby
.where(:resource_type => authentication_mixin_relation)
```

However, when I tried with the following diff:

```diff
diff --git a/app/models/mixins/authentication_mixin.rb b/app/models/mixins/authentication_mixin.rb
index e38908af97..f9775d4f37 100644
--- a/app/models/mixins/authentication_mixin.rb
+++ b/app/models/mixins/authentication_mixin.rb
@@ -36,7 +36,8 @@ module AuthenticationMixin
       authentication_mixin_relation = name
 
       proc do
-        where('"authentications"."resource_type" = ?', authentication_mixin_relation)
+        # where('"authentications"."resource_type" = ?', authentication_mixin_relation)
+        where(:authentications => {:resource_type => authentication_mixin_relation})
           .order(Authentication::STATUS_SEVERITY_AREL.desc)
       end
     end
```

I got the following error:

<details>
<summary><code>$ bundle exec rspec spec/models/mixins/authentication_mixin_spec.rb:788</code></summary>

```console
$ bundle exec rspec spec/models/mixins/authentication_mixin_spec.rb:788
** ManageIQ master, codename: Kasparov
Run options: include {:locations=>{"./spec/models/mixins/authentication_mixin_spec.rb"=>[788]}}

Randomized with seed 58849
FF.

Failures:

  1) AuthenticationMixin#authentication_status with an unrelated and existing valid authentication is nil with sql
     Failure/Error: query.first["extra"]

     ActiveRecord::StatementInvalid:
       PG::DatatypeMismatch: ERROR:  argument of LIMIT must be type bigint, not type text
       LINE 1: ...") = 'invalid' THEN 3 ELSE -1 END DESC LIMIT 1) ASC LIMIT $1
                                                                            ^
       : SELECT  "hosts"."id", (SELECT  "authentications"."status" FROM "authentications" WHERE "authentications"."resource_type" = $1 AND "authentications"."resource_id" = "hosts"."id" ORDER BY CASE WHEN LOWER("authentications"."status") = '' THEN -1 WHEN LOWER("authentications"."status") = 'valid' THEN 0 WHEN LOWER("authentications"."status") = 'none' THEN 1 WHEN LOWER("authentications"."status") = 'incomplete' THEN 1 WHEN LOWER("authentications"."status") = 'error' THEN 2 WHEN LOWER("authentications"."status") = 'unreachable' THEN 2 WHEN LOWER("authentications"."status") = 'invalid' THEN 3 ELSE -1 END DESC LIMIT 1) AS extra FROM "hosts" ORDER BY (SELECT  "authentications"."status" FROM "authentications" WHERE "authentications"."resource_type" = $1 AND "authentications"."resource_id" = "hosts"."id" ORDER BY CASE WHEN LOWER("authentications"."status") = '' THEN -1 WHEN LOWER("authentications"."status") = 'valid' THEN 0 WHEN LOWER("authentications"."status") = 'none' THEN 1 WHEN LOWER("authentications"."status") = 'incomplete' THEN 1 WHEN LOWER("authentications"."status") = 'error' THEN 2 WHEN LOWER("authentications"."status") = 'unreachable' THEN 2 WHEN LOWER("authentications"."status") = 'invalid' THEN 3 ELSE -1 END DESC LIMIT 1) ASC LIMIT $1
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:611:in `exec_params'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:611:in `block (2 levels) in exec_no_cache'
     # ~/.gem/ruby/2.6.3/gems/activesupport-5.2.4.3/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
     # ~/.gem/ruby/2.6.3/gems/activesupport-5.2.4.3/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
     # ~/.gem/ruby/2.6.3/gems/activesupport-5.2.4.3/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:610:in `block in exec_no_cache'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:581:in `block (2 levels) in log'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:580:in `block in log'
     # ~/.gem/ruby/2.6.3/gems/activesupport-5.2.4.3/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:571:in `log'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:609:in `exec_no_cache'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:598:in `execute_and_clear'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/database_statements.rb:84:in `exec_query'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/database_statements.rb:478:in `select'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/database_statements.rb:70:in `select_all'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/query_cache.rb:106:in `select_all'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/querying.rb:41:in `find_by_sql'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:560:in `block in exec_queries'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:584:in `skip_query_cache_if_necessary'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:547:in `exec_queries'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:422:in `load'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:200:in `records'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:195:in `to_ary'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:532:in `find_nth_with_limit'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:517:in `find_nth'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:125:in `first'
     # ./spec/support/arel_helper.rb:15:in `virtual_column_sql_value'
     # ./spec/models/mixins/authentication_mixin_spec.rb:796:in `block (4 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # PG::DatatypeMismatch:
     #   ERROR:  argument of LIMIT must be type bigint, not type text
     #   LINE 1: ...") = 'invalid' THEN 3 ELSE -1 END DESC LIMIT 1) ASC LIMIT $1
     #                                                                        ^
     #   ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:611:in `exec_params'

  2) AuthenticationMixin#authentication_status with an unrelated and existing valid authentication is 'None' when accessed via ruby, but fetched via sql
     Failure/Error: fetched_host = Host.select(:id, :authentication_status).first

     ActiveRecord::StatementInvalid:
       PG::DatatypeMismatch: ERROR:  argument of LIMIT must be type bigint, not type text
       LINE 1: ...tion_status" FROM "hosts" ORDER BY "hosts"."id" ASC LIMIT $1
                                                                            ^
       : SELECT  "hosts"."id", (SELECT  "authentications"."status" FROM "authentications" WHERE "authentications"."resource_type" = $1 AND "authentications"."resource_id" = "hosts"."id" ORDER BY CASE WHEN LOWER("authentications"."status") = '' THEN -1 WHEN LOWER("authentications"."status") = 'valid' THEN 0 WHEN LOWER("authentications"."status") = 'none' THEN 1 WHEN LOWER("authentications"."status") = 'incomplete' THEN 1 WHEN LOWER("authentications"."status") = 'error' THEN 2 WHEN LOWER("authentications"."status") = 'unreachable' THEN 2 WHEN LOWER("authentications"."status") = 'invalid' THEN 3 ELSE -1 END DESC LIMIT 1) AS "authentication_status" FROM "hosts" ORDER BY "hosts"."id" ASC LIMIT $1
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:611:in `exec_params'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:611:in `block (2 levels) in exec_no_cache'
     # ~/.gem/ruby/2.6.3/gems/activesupport-5.2.4.3/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
     # ~/.gem/ruby/2.6.3/gems/activesupport-5.2.4.3/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
     # ~/.gem/ruby/2.6.3/gems/activesupport-5.2.4.3/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:610:in `block in exec_no_cache'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:581:in `block (2 levels) in log'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:580:in `block in log'
     # ~/.gem/ruby/2.6.3/gems/activesupport-5.2.4.3/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:571:in `log'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:609:in `exec_no_cache'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:598:in `execute_and_clear'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/database_statements.rb:84:in `exec_query'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/database_statements.rb:478:in `select'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/database_statements.rb:70:in `select_all'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/query_cache.rb:106:in `select_all'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/querying.rb:41:in `find_by_sql'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:560:in `block in exec_queries'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:584:in `skip_query_cache_if_necessary'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:547:in `exec_queries'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:422:in `load'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:200:in `records'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:195:in `to_ary'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:532:in `find_nth_with_limit'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:517:in `find_nth'
     # ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:125:in `first'
     # ./spec/models/mixins/authentication_mixin_spec.rb:804:in `block (4 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # PG::DatatypeMismatch:
     #   ERROR:  argument of LIMIT must be type bigint, not type text
     #   LINE 1: ...tion_status" FROM "hosts" ORDER BY "hosts"."id" ASC LIMIT $1
     #                                                                        ^
     #   ~/.gem/ruby/2.6.3/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:611:in `exec_params'

Finished in 2.83 seconds (files took 8.93 seconds to load)
3 examples, 2 failures

Failed examples:

rspec ./spec/models/mixins/authentication_mixin_spec.rb:795 # AuthenticationMixin#authentication_status with an unrelated and existing valid authentication is nil with sql
rspec ./spec/models/mixins/authentication_mixin_spec.rb:803 # AuthenticationMixin#authentication_status with an unrelated and existing valid authentication is 'None' when accessed via ruby, but fetched via sql

Randomized with seed 58849
```

</details>